Ability to set content of static pages via Site Configuration

macornwell opened the pull request:
This PR allows the ability to set the header `<h1>` and content sections of the static template pages in two different ways.
1. Using the pageheader and pagecontent block names as a site wide template.
2. Setting SiteConfiguration settings for the header and/or the content for each page.

**Dependencies**: None

**Screenshots**:
![screenshot](https://user-images.githubusercontent.com/2694231/31840315-6c421d86-b5aa-11e7-8ad5-ecf3bb428fa9.jpg)


**Sandbox URL**: https://pr16297.sandbox.opencraft.hosting

**Merge deadline**: "None" 

**Testing instructions**:
1. Open the django admin for https://pr16297.sandbox.opencraft.hosting
2. Login with the typical staff credentials.
3. Add a new SiteConfiguration model.
4. Select the relevant site (could be example.com)
5. For the value insert the following:
`{"static_template_about_header":"Unique About Header", "static_template_about_content":"The content of the about page."}`
6. Navigate to https://pr16297.sandbox.opencraft.hosting/about
7. Ensure that the header and content have changed as expected.

**Author notes and concerns**:
1. The format for modifying the SiteConfiguration is as follows.
For the header: `static_template_htmlpagename_header`
For the content: `static_template_htmlpagename_content`
If the page contains a dash '-' it is converted to an underscore '_'.
The ".html" is also removed as well.
Example: `static_template_server_error_content`
2. The content allows html so that a more complex content section can be provided such as `<p>` for separating out paragraphs or `<img>` for images.

**Reviewers**
- [ ] smarnach
- [ ] edX reviewer[s] TBD

OpenCraft ticket: OC-3225

edx-webhook commented:
Thanks for the pull request, @macornwell! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=16297&repo=edx%2Fedx-platform"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

macornwell commented:
jenkins run quality

smarnach commented on lms/djangoapps/static_template_view/tests/test_views.py:
```diff
@@ -26,6 +37,20 @@ def test_about(self):
+        test_header_key = 'static_template_about_header'
+        test_content_key = 'static_template_about_content'
+        response = None
+        with with_site_configuration_context(configuration={test_header_key: test_header,test_content_key: test_content}):
```
You still have a PEP8 violation in this line (which causes the quality Jenkins task to fail): missing space after the comma.

smarnach commented on lms/djangoapps/static_template_view/tests/test_views.py:
```diff
@@ -5,14 +5,25 @@
+        Sets up the test.
+        """
+        from django.test.utils import setup_test_environment
+        setup_test_environment()
```
According to the [documentation](
https://github.com/django/django/blob/e3cf8d2e94fe5f0b6d41774edbab0a9800b397ff/django/test/utils.py#L99-L105) of this function, this is for global test setup and should be run once by the test runner, and not for individual tests.  What's the idea of calling it here?

macornwell replied:
When running the tests locally last week it said it needed it. Deleting it and running it now, it doesn't. Deleted the method.

smarnach commented on lms/djangoapps/static_template_view/tests/test_views.py:
```diff
@@ -5,14 +5,25 @@
         resp = self.client.get(reverse(view_name))
         self.assertEqual(resp.status_code, 200)
         self.assertEqual(resp['Content-Type'], mimetype)
+        return resp
```
Why this change?

macornwell replied:
It was originally because I was using it as a setup method, but I am no longer using it. Line deleted.

smarnach commented on lms/djangoapps/static_template_view/views.py:
```diff
@@ -50,7 +52,12 @@ def render(request, template):
+        # Format Examples: static_template_about_header
+        configuration_base = 'static_template_' + template.replace('.html', '').replace('-', '_')
+        context['page_header'] = configuration_helpers.get_value(configuration_base + '_header')
+        context['page_content'] = SafeUnicode(configuration_helpers.get_value(configuration_base + '_content'))
```
Wouldn't we want to allow HTML markup in the header as well?  At least HTML entities like `&auml;` should be allowed, so we should add `SafeUnicode` there as well.

macornwell replied:
I agree. Added.

macornwell replied:
After looking at and trying a few different ways to collapse these down, I haven't been able to find something that both works, and meaningfully reduces the duplication. It just added more code, another new template, and about ~40% less templates.

One other possibility is deleting all of the more "generic" template files and just creating a single generic template that you pass in the Title and Header in into the view.

But I felt that was changing the functionality a bit too far as changing the SiteConfiguration would be the only way.

macornwell commented:
@smarnach is the sandbox running with the current version? I re-pulled down this code and got it running and the html works as expected. I'm in the process of writing some unit tests to automatically check for the html functionality.

smarnach commented:
@macornwell Looks all good to me now, :+1:.  Thanks for seeing this through!

Can you please update the links in the testing instructions to the actual sandbox?  This will make it easier for people at edX to test it.  Please also mention that the usual staff account can be used for testing.

- [x] I tested this: (describe what you tested)
- [x] I read through the code

edx-webhook commented:
Thanks for the pull request, @macornwell! I've created [OSPR-1984](https://openedx.atlassian.net/browse/OSPR-1984) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

nedbat commented on lms/djangoapps/static_template_view/tests/test_views.py:
```diff
@@ -26,6 +32,64 @@ def test_about(self):
+    def test_about_with_site_configuration(self):
+        """
+        Test the about view with the header and content set in SiteConfiguration.
+        """
```
BTW, there is no need for docstrings in test_* methods, and they tend to be kind of obvious anyway.  Feel free to use comments if you need to explain something.

nedbat commented:
@macornwell Can you help us understand why you are making this change? Maybe I'm wrong, but I thought people just edited the templates if they wanted to change the contents.  I guess that's not a great way to do it....

macornwell commented:
@nedbat  The idea behind the PR is that consumers long term would be able to make modifications to the site through the admin interface rather than the code itself.

nedbat commented:
I'm trying to understand the motivation to take on this work.  Is this something OpenCraft will use, or your customers have asked for?  These pages are all vestiges of the original edx.org website.

smarnach commented:
@nedbat We are currently implementing a few features to make it possible to configure simple themes via the Ansible vars.  In the standard Open edX theme, many of the static pages are accessible via links, so they need to contain at least some content.  Having a terms of service page linked on the registration form that simply states "This page is intentionally left empty" isn't really useful, so these static pages need to be customised for virtually every Open edX installation.

Being able to do basic theme customizations via the Ansible vars is useful for automation.  It's much easier to dump a bunch of additional variables into the configuration than it is to programmatically create new branches in a Github comprehensinve theme repo.

There would be several alternatives to the approach chosen here.  I wasn't involved in the original design, so I internally asked for further information.

nedbat commented on lms/djangoapps/static_template_view/tests/test_views.py:
```diff
@@ -26,6 +32,64 @@ def test_about(self):
+        configuration = {test_header_key: test_header, test_content_key: test_content}
+        with with_site_configuration_context(configuration=configuration):
+            response = self.client.get(reverse("about"))
+        self.assertIn(test_content, response.content)
```
Why aren't you checking that the header is in the content?  The test is about the header.  Also, seems like this test and the next one could be combined.

nedbat commented on lms/djangoapps/static_template_view/tests/test_views.py:
```diff
@@ -26,6 +32,64 @@ def test_about(self):
+        with with_site_configuration_context(configuration=configuration):
+            response = self.client.get(reverse("about"))
+        escaped_content = escape(test_content)
+        self.assertNotIn(str(escaped_content), response.content)
```
This test method seems unnecessary: the previous test confirms that the HTML content is in the response verbatim, so why run another test just to check that the escaped content isn't in the response.  If you want to explicitly check this condition, just add the assertNotIn to test_about_with_site_configuration_and_html_in_content

macornwell replied:
This is a good question. In hindsight I am not 100% sure why I separated them. The only thing I can think of is that the unit test as a single unit is testing two different things, but in practicality both couldn't be happening at the same time. (Value is escaped and the value isn't escaped).

nedbat commented on lms/djangoapps/static_template_view/views.py:
```diff
@@ -50,7 +52,16 @@ def render(request, template):
+        if page_header:
+            context['page_header'] = SafeUnicode(page_header)
+        if page_content:
+            context['page_content'] = SafeUnicode(page_content)
```
I'm a little confused why this is the first use of `SafeUnicode` or `SafeText` in edx-platform.  How has this problem been solved in other places?

nedbat replied:
`mark_safe` is used a few times in edx-platform, and will accomplish the same thing (but more flexibly).  I'm still a little confused why something like this doesn't appear more often.

macornwell replied:
I noticed the same thing as I am looking through it. I chose SafeUnicode because its the standard Django approach. As far as why its not entirely sure why there isn't more. I'm switching it over now.

macornwell commented:
@nedbat I made the fixes and switched to mark_safe. I changed the unit tests so that they use Unicode strings which is what is expected inside the application. I ran into an issue where the code worked and displayed properly in LMS, but failed unit tests.

nedbat commented:
@macornwell Thanks.  There's a pep8 violation to fix (./lms/djangoapps/static_template_view/tests/test_views.py:66:5: E303 too many blank lines (2))

You should squash the commits.
The bok-choy failure I think has been fixed on master, so when you rebase and squash, we can see the tests pass again.

mtyaka commented:
Thanks for reviewing this @nedbat! To further clarify the motivation behind this change:

Most of OpenCraft clients are using at least some of these static pages. There is currently no way to set the content of these pages without overriding the templates in a custom theme. Since maintaining separate themes for each client is not very efficient, we've been working on [automating some aspects of instance customization](https://github.com/edx/configuration/pull/4052). We decided to use site configuration variables because they are generic (no need for custom logic to automate static page template overrides) and because a similar approach is already used to set the [homepage overlay html](https://github.com/edx/edx-platform/blob/a904324f6dd21056725a6d9f9fc834660b920efe/lms/templates/index.html#L17-L24), which is another part of the system that almost every client wants customized.

macornwell commented:
@smarnach  @nedbat  @mtyaka  where are we at with this Pull Request? Is there something more that needs to be done or discussed?

smarnach commented:
@macornwell You can rebase on master, since there are conflicts with upstream again.  This will also trigger the tests to run again on current master, so we will be able to see whether bokchoy has been fixed.

nedbat commented:
jenkins run bokchoy

nedbat commented:
jenkins run python

macornwell commented:
jenkins run python

macornwell commented:
@nedbat I've got the unit tests fixed on this PR and its ready for another review.

macornwell commented:
jenkins run bokchoy

nedbat commented:
I am going to be away for a week, but I promise I will finish this pull request when I get back.

nedbat commented:
I'm wondering about the <%block>'s you have: do you plan to override these block in inherited pages?

Also, this could be shortened:
```
            <%block name="pageheader">
                % if page_header:
                    ${page_header}
                % else:
                    ${_("Media Kit")}
                % endif
            </%block>
```
to
```
<%block name="pageheader">${page_header or _("Media Kit")}</%block>
```

nedbat commented:
jenkins run all

nedbat commented:
jenkins run all

nedbat commented:
@macornwell The tests won't pass until this is rebased onto master.

macornwell commented:
@nedbat  looks like everything is green now.

nedbat commented:
@macornwell thanks, please squash the commits.  What is the answer to the question about the blocks?  And do you have an opinion about the more compact code?

macornwell commented:
@nedbat  just tested it out. Looks like a solid change. 
I'll get on converting these templates, squash and then send this up.

nedbat commented on lms/templates/static_templates/404.html:
```diff
@@ -10,11 +10,21 @@
-        email=HTML('<a href="mailto:{email}">{email}</a>').format(email=Text(static.get_tech_support_email_address()))
-      )}</p>
+        <h1>
+            <%block name="pageheader">${page_header or _("Page Not Found")}</%block>
```
There are test failures, because the old code had "Page not found", and now we have "Page Not Found".

nedbat commented:
The test failures seem real.  One is due to the case of "Page Not Found", but there are also Mako errors.  I know this has been a long process, we there will have to get fixed.

nedbat commented:
@macornwell Will you be able to work on these test failures?

macornwell commented:
@nedbat  I will not be able to.

bradenmacdonald commented:
@nedbat Sorry for the delay; I'm working to find someone else from my team to fix this and see this PR through to merging ASAP.

tomaszgy commented:
jenkins run python
jenkins run django-upgrade/python

tomaszgy commented:
@nedbat I've added a small conditional check in `embargo.html` and it seems that the tests pass now :)

edx-webhook commented:
@macornwell 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/16297&entry.1484655318=https://github.com/macornwell&entry.752974735=2017-10-20+18:01&entry.1917517419=2018-03-28+18:28&entry.2133058324=Yes) so we can improve your experience in the future.

tomaszgy commented:
Thanks @nedbat! :)
